Add SecurityTokenSignatureKeyNotFoundWithValidationErrorsException

keegan-caruso opened the pull request:
SecurityTokenSignatureKeyNotFoundWithValidationErrorsException represents
SecurityTokenSignatureKeyNotFoundException but with validation errors
such that a different signal is sent to the layer above

keegan-caruso committed:
Add SecurityTokenSignatureKeyNotFoundWithValidationErrorsException

SecurityTokenSignatureKeyNotFoundWithValidationErrorsException represents
SecurityTokenSignatureKeyNotFoundException but with validation errors
such that a different signal is sent to the layer above

GeoK commented on test/Microsoft.IdentityModel.TestUtils/Default.cs:
```diff
@@ -279,7 +279,7 @@ public static DateTime Expires
         public static string ExpiresString
         {
-            get => "2021-03-17T18:33:37.080Z";
+            get => "2025-03-17T18:33:37.080Z";
```
Wow, what timing! This date was created 3y ago.

GeoK commented on test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs:
```diff
@@ -2687,6 +2687,96 @@ public static TheoryData<JWEDecompressionTheoryData> JWEDecompressionTheoryData(
+                    {
+                        IssuerSigningKey = Default.SymmetricSigningKey,
+                    },
+                    ExpectedException = ExpectedException.SecurityTokenSignatureKeyNotFoundWithValidationErrorsException("IDX10516:")
```
Issuer mismatch? Consider adding a testId for clarity.

GeoK commented on test/Microsoft.IdentityModel.Tokens.Saml.Tests/SamlSecurityTokenHandlerTests.cs:
```diff
@@ -1285,6 +1285,96 @@ public static TheoryData<CreateTokenTheoryData> CreateSamlTokenUsingTokenDescrip
+                    ValidationParameters = new TokenValidationParameters
+                    {
+                        IssuerSigningKey = Default.SymmetricSigningKey,
+                        ValidateIssuer = false,
```
Any reason not to set 'ValidIssuer'? (here and on other places)

GeoK commented on src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.cs:
```diff
@@ -1104,7 +1104,13 @@ private SamlSecurityToken ValidateSignature(SamlSecurityToken samlToken, string
-                throw LogHelper.LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(LogHelper.FormatInvariant(TokenLogMessages.IDX10513, samlToken.Assertion.Signature.KeyInfo, exceptionStrings, samlToken)));
+                InternalValidators.ValidateLifetimeAndIssuerAfterSignatureNotValidatedSaml(
+                    samlToken,
+                    samlToken.Assertion.Conditions.NotBefore,
```
Check for null Conditions? (for both saml handlers)

GeoK commented on src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:
```diff
@@ -1017,10 +1017,19 @@ protected virtual JwtSecurityToken ValidateSignature(string token, TokenValidati
+                DateTime? expires = (jwtToken.Payload.Exp == null) ? null : new DateTime?(jwtToken.ValidTo);
+                DateTime? notBefore = (jwtToken.Payload.Nbf == null) ? null : new DateTime?(jwtToken.ValidFrom);
+
+                InternalValidators.ValidateLifetimeAndIssuerAfterSignatureNotValidatedJwt(
```
ValidateSignature/ValidateToken are missing the description about the new exception type. That could be a good place to make a point about differences between SecurityTokenSignatureKeyNotFoundException and SecurityTokenSignatureKeyNotFoundWithValidationErrorsException.

GeoK commented on src/Microsoft.IdentityModel.Tokens/Exceptions/SecurityTokenSignatureKeyNotFoundWithValidationErrorsException.cs:
```diff
@@ -0,0 +1,136 @@
+    /// In addition, when validation errors exist over the security token.
+    /// </summary>
+    [Serializable]
+    public class SecurityTokenSignatureKeyNotFoundWithValidationErrorsException
```
How can we better communicate the intent behind SecurityTokenSignatureKeyNotFoundException? Should we update the summary to hint towards metadata refresh?

brentschmaltz replied:
How about: SecurityTokenUnableToValidate and then add a public external property that describes the error.

GeoK commented:
Conceptually looks good.

keegan-caruso committed:
Add exception message to JwtSecurityTokenHandler.ValidateSignature

SecurityTokenSignatureKeyNotFoundWithValidationErrorsException is not a
signal for metadata refresh

keegan-caruso committed:
Update reference tokens to not be expired

keegan-caruso committed:
Update tests for clarity

keegan-caruso committed:
update tests to generate AT that is validated

keegan-caruso committed:
SecurityTokenSignatureKeyNotFoundWithValidationErrorsException message

Update message to indicate it is not used for metadata refresh

keegan-caruso committed:
Null checks for saml handlers

keegan-caruso committed:
Test fixes

keegan-caruso committed:
Use SecurityTokenUnableToValidateException

Remove SecurityTokenSignatureKeyNotFoundWithValidationErrorsException and
use SecurityTokenUnableToValidateException instead, allowing the single
exception type to express different types of validation failures

Add ValidationFailure as a bit field enum

keegan-caruso committed:
fix header

GeoK commented on src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.cs:
```diff
@@ -1104,7 +1104,14 @@ private SamlSecurityToken ValidateSignature(SamlSecurityToken samlToken, string
                     throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10514, keysAttempted, samlToken.Assertion.Signature.KeyInfo, exceptionStrings, samlToken)));
 
-                throw LogHelper.LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(LogHelper.FormatInvariant(TokenLogMessages.IDX10513, samlToken.Assertion.Signature.KeyInfo, exceptionStrings, samlToken)));
+                if (samlToken.Assertion.Conditions != null && samlToken.Assertion.Signature.KeyInfo != null)
```
samlToken.Assertion.Signature.KeyInfo should be already verified at this point.

Could we do something like:
samlToken.Assertion.Conditions?.NotBefore
samlToken.Assertion.Conditions?.NotOnOrAfter

and call into InternalValidators?

keegan-caruso replied:
Will remove the check on Signature.KeyInfo.

Saml1 doesn't use nullable datetimes, so can only check the conditions object for null.

For saml2, I think we pass along null NotBefore and NotOnOrAfter and let that contribute to the decision on lifetime validity. Does that sound right?

keegan-caruso committed:
Remove check on KeyInfo != null

mafurman commented on src/Microsoft.IdentityModel.Tokens/Exceptions/SecurityTokenUnableToValidateException.cs:
```diff
@@ -0,0 +1,121 @@
+{
+    /// <summary>
+    /// This exception is thrown when a security token contained a key identifier but the key was not found by the runtime.
+    /// In addition, when validation errors exist over the security token. This exception is not intended to be used as a signal
```
I'd suggest rephrasing this as "key was not found by the runtime AND when validation errors...." to make it clearer both conditions need to be met (assuming I'm understanding this correctly).

mafurman commented on src/Microsoft.IdentityModel.Tokens/InternalValidators.cs:
```diff
@@ -0,0 +1,133 @@
+    {
+        /// <summary>
+        /// Called after signature validity has already been determined
+        /// </summary>
```
Would it be more correct to say "called after signature validation has already failed"? We always throw here, right?

brentschmaltz replied:
I agree with this, we call these when we couldn't validate the signature.

mafurman commented on src/Microsoft.IdentityModel.Tokens/Exceptions/ValidationFailure.cs:
```diff
@@ -0,0 +1,48 @@
+        /// Indicates that the issuer was invalid
+        /// </summary>
+        InvalidIssuer = 0b_0000_0000_0000_0010, // 2
+    }
```
I might be misunderstanding something.... do we need an enum that represents when both are invalid?

keegan-caruso replied:
The idea of using a bitflag: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/enum#enumeration-types-as-bit-flags

That way we can represent a combination. And then the end user will do .HasFlag() to determine if a particular value is set.

If you think that is confusing, we can use distinct enums values

brentschmaltz replied:
If we have a bitmap, then would we need a helper, cause we can't use ==.

brentschmaltz replied:
We decided a bitmap is fine.

keegan-caruso committed:
Improve comments

brentschmaltz commented on test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs:
```diff
@@ -2687,6 +2703,107 @@ public static TheoryData<JWEDecompressionTheoryData> JWEDecompressionTheoryData(
+            try
+            {
+                var handler = new JsonWebTokenHandler();
+
```
nit: do we need all these empty lines?

keegan-caruso replied:
removed

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/InternalValidators.cs:
```diff
@@ -0,0 +1,144 @@
+
+            if (validLifetime && validIssuer)
+                throw LogHelper.LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(LogHelper.FormatInvariant(TokenLogMessages.IDX10501, kid, exceptionStrings, securityToken)));
+            else
```
code could read slightly simplier.
if (validLifetime && validIssuer) ...
else if (validLifetime)
    throw InvalidIssuer ...
else if (validIssuer)
    throw InvalidLifetime ...

Similar in saml

keegan-caruso replied:
Because of the bitfield enum I don't think we can reduce it that way. it would become roughly

if (validLifetime && validIssuer) => Key not found
if (validLifetime && !validIssuer) => UnableToValidate(invalid issuer)
if (!validLifetime && validIssuer) => UnableToValidate(invalid lifetime)
else => UnableToValdiate(invalid issuer & invalid lifetime)

I think

keegan-caruso replied:
I worked through this and it looked more complicated to me

keegan-caruso committed:
remove unneeded whitespace
